domain/machine/modelmigration/import.go

```diff
@@ -55,8 +60,46 @@ func (i *importOperation) Setup(scope modelmigration.Scope) error {
 func (i *importOperation) Execute(ctx context.Context, model description.Model) error {
 	for _, m := range model.Machines() {
 		// We need skeleton machines in dqlite.
-		if _, err := i.service.CreateMachine(ctx, machine.Name(m.Id())); err != nil {
-			return fmt.Errorf("importing machine %q: %w", m.Id(), err)
+		machineUUID, err := i.service.CreateMachine(ctx, machine.Name(m.Id()))
+		if err != nil {
+			return errors.Annotatef(err, "importing machine %q", m.Id())
+		}
+
+		// Import the machine's cloud instance.
+		cloudInstance := m.Instance()
+		if cloudInstance != nil {
+			arch := cloudInstance.Architecture()
+			mem := cloudInstance.Memory()
+			rootDisk := cloudInstance.RootDisk()
+			rootDiskSource := cloudInstance.RootDiskSource()
+			cpuCores := cloudInstance.CpuCores()
+			cpuPower := cloudInstance.CpuPower()
+			tags := cloudInstance.Tags()
+			virtType := cloudInstance.VirtType()
+			hardwareCharacteristics := &instance.HardwareCharacteristics{
+				Arch:           &arch,
+				Mem:            &mem,
+				RootDisk:       &rootDisk,
+				RootDiskSource: &rootDiskSource,
+				CpuCores:       &cpuCores,
+				CpuPower:       &cpuPower,
+				Tags:           &tags,
+				VirtType:       &virtType,
+			}
+			// Only add the availability zone if it is not empty. It can be empty
+			// because we are deserializing from a description.Model and we
+			// lose the pointer reference since it returns a string.
```

Comment:
We should do it to all the other values above as well. That way we ensure we have clean and __valid__ data entering into the domain.

Reply:
**suggestion:** Use a small generic function. 

```go
func nilZeroPtr[T comparable](v T) *T {
    var zero T
    if v == zero {
        return nil
    }
    return &v
}
```

(Ok maybe with a better name :) )

Reply:
Sure, I like your idea @gfouillet. Done in 9b4a2b06aa357bd3119ef601318c0f923c583ff7.